cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -36,36 +43,34 @@
 public class BasicWorkflowToken implements WorkflowToken, Serializable {
 
   private static final long serialVersionUID = -1173500180640174909L;
+  private static final Object sharedLock = new Object();
 
   private Map<String, Map<String, Long>> mapReduceCounters;
   private final Map<Scope, Map<String, List<NodeValue>>> tokenValueMap = new EnumMap<>(Scope.class);
   private String nodeName;
+  private Set<String> nodeFilterSet;
 
   public BasicWorkflowToken() {
     for (Scope scope : Scope.values()) {
-      tokenValueMap.put(scope, new HashMap<String, List<NodeValue>>());
+      tokenValueMap.put(scope, new ConcurrentHashMap<String, List<NodeValue>>());
     }
   }
 
   private BasicWorkflowToken(BasicWorkflowToken other) {
-    for (Map.Entry<Scope, Map<String, List<NodeValue>>> entry : other.tokenValueMap.entrySet()) {
-      Map<String, List<NodeValue>> tokenValueMapForScope = new HashMap<>();
-      for (Map.Entry<String, List<NodeValue>> valueEntry : entry.getValue().entrySet()) {
-        tokenValueMapForScope.put(valueEntry.getKey(), Lists.newArrayList(valueEntry.getValue()));
-      }
-
-      this.tokenValueMap.put(entry.getKey(), tokenValueMapForScope);
+    // Same instance of the tokenValueMap is shared across all the
+    // nodes in the Workflow, so that values can be put concurrently.
```

Comment:
Revised the code. So as per the current logic, we always make a deep copy of the WorkflowToken before starting the execution of the ACTION/CONDITION. Put operations happen on the copied token. The copied token is merged back to the original token after the execution of the node completes.